Review

Karim-Mane opened the pull request:
* **Please check if the PR fulfills these requirements**

- [x] I have read the CONTRIBUTING guidelines
- [x] A new item has been added to `NEWS.md`
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)
- [x] Checks have been run locally and pass

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

This is a full package review PR. Changes and new features are mentioned in the `NEWS.md` file.

joshwlambert reviewed (commented):
@Karim-Mane thanks for opening the {cleanepi} review. Really nice work on the package! 🚀  

The use of {cli} for messages and errors has definitely improved the user experience. 

I didn't have time to go through the whole package, but here are my comments and several questions. If anything is unclear please let me know and I'm happy to discuss and clarify.

- The `ISSUE_TEMPLATE` and `PULL_REQUEST_TEMPLATE` folders can be deleted as this is handled by the .github repository in the epiverse-trace GitHub organisation
- `check_date_sequence()` I recommend making the function documentation title more precise, e.g. "Check whether the sequence of date-events is in chronological order".
- _Personal preference_: I would use monospace or angular brackets or both to denote classes. For example in `check_date_sequence()` the "@param data The input data frame or linelist" could be "@param data The input data frame or `linelist`" or "@param data The input data frame or `<linelist>`"
- Not a request but just making you aware in case you decided it would be useful. In `clean_data()` you could accept [dynamic-dots](https://rlang.r-lib.org/reference/dyn-dots.html). This was suggested by Hugo for some functions in packages I maintain. 
- _Question_: I haven't check this myself, do you know if the functions that accept a `data.frame` also work correctly on `<tibble>` or `<data.table>`?
- There seem to be a few functions that use `data` as the first argument and have similar if not the same documentation. If the argument documentation is indeed identical and duplicated I recommend using `@inheritParam` and only documenting the argument in a single place (probably `clean_data()` as it's the central exported function).
- _Question_: Is there any functionality that uses the `report` attribute to automate other cleaning functions, or is it just meant for users to inspect? I might be nice to consider if the report can be used by some cleaning functions. 
- _Question_: should {naniar} be a suggested dependency as it is used in `data-raw/`?
- In the `DESCRIPTION` you're using a development version of {roxygen2}, `RoxygenNote: 7.3.2.9000`, maybe better to re-render the documentation with a stable release of {roxygen}. I don't think it will cause any issues but just in case as CRAN usually don't like development versions of packages.
- The {cleanepi} package started development in 2022, the copyright in the `LICENSE` files can be 2022-2025.
- It would be good if you could back-date the `NEWS.md` and provide some information on previous releases.
- I would remove "curating" from the package `Description:` in the `DESCRIPTION` file.
- Add PR or issue tags to `NEWS.md` bullet points currently missing them, if possible.
- I'm a bit confused by the `add_to_dictionary()` function, I'm not sure how much value this function adds over a user just calling `rbind()` to add to a dictionary. I've seen the description in the `cleanepi.Rmd` vignette, but I would appreciate a bit more explanation to where you see this function being used. 
- Running the `add_to_report()` example the output of `clean_data()` is a `<data.table>`, I would expect it to be a `<data.frame>` as that is the input type. I can't see anywhere where {data.table} is called so I assume this is a dependency doing this conversion. Can you check to see if this is also the case for you, and if so it would be good to return the same tabular data type that the user inputs.
- _Question_: in `add_to_report()` is it possible to add a list element to the `report` attribute in place rather than extracting it, modifying it and then replacing it. I would only be a marginal gain either way. 
- Would be good to add to the documentation of `convert_to_numeric()` that it is a wrapper around `numberize::numberize()`
- `test_df1.rds` has a typo in the column name, `recruted_on_day` should be `recruited_on_day`.
- I recommend renaming the `ref_date` of the `convert_numeric_to_date()` function to `origin` and inheriting the documentation from `as.Date()`. ***Edit*** I see this isn't exactly the same as `origin` as it interactive with the `forward` argument. I find this a bit confusing, so perhaps ignore my original point.
- Recently (R 4.3.0) I think the behaviour of `as.Date.numeric()` changed, it might be worth seeing how this affects `convert_numeric_to_date()`
- _Question_: I'm not sure how much you want to automate because of false positives/negatives but it might be worth having some amount of automated detecting of column types and coercing for `convert_*()` functions. For example if I have a line list with date columns that are `numeric`, is it beneficial to not specify the `target_columns` argument and instead find all numeric columns that could be converted to dates? It would add complexity to the functions, and I'm not saying this is definitely the way to go, but might be something to consider.
- _Question_: how does `correct_subject_ids()` differ from `clean_using_dictionary()`, they both seem to clean a `<data.frame>` using a user-specified `<data.frame>`, albeit with slightly different structures. Could these functions be merged? It could be preferable to channel all `<data.frame>` based cleaning through `clean_using_dictionary()`?

joshwlambert commented:
In `check_subject_ids()`, it was unclear to me on first reading whether the `prefix` and `suffix` arguments are identifying the correct or incorrect ID formats. It would be good to clarify the function documentation to make this clear.

Karim-Mane commented:
Thank you @joshwlambert for your review. I will address the issues to spotted out and respond to the comments below.

Karim-Mane commented:
> The ISSUE_TEMPLATE and PULL_REQUEST_TEMPLATE folders can be deleted as this is handled by the .github repository in the epiverse-trace GitHub organisation

I have accounted for this.

Karim-Mane commented:
> `check_date_sequence()` I recommend making the function documentation title more precise, e.g. "Check whether the sequence of date-events is in chronological order".

Thanks for the suggestion. I have accounted for this in commit `baa601a`.

Karim-Mane commented:
> Personal preference: I would use monospace or angular brackets or both to denote classes. For example in check_date_sequence() the "@param data The input data frame or linelist" could be "@param data The input data frame or linelist" or "@param data The input data frame or <linelist>.

This is a good suggestion. I am already using the <linelist> style in other parts of my work. This has been accounted for in commit `326d24b`.

Karim-Mane commented:
> Not a request but just making you aware in case you decided it would be useful. In clean_data() you could accept [dynamic-dots](https://rlang.r-lib.org/reference/dyn-dots.html). This was suggested by Hugo for some functions in packages I maintain.

Thanks @joshwlambert for the suggestion. I will stick with the current code, and explore this for the next versions. Will add it in the TODO list of the package project

Karim-Mane commented:
> Question: I haven't check this myself, do you know if the functions that accept a data.frame also work correctly on <tibble> or <data.table>?

Yes - I have successfully executed the functions on both <tibble> and <data.table> objects. 

Please create issue if you find any function that do not work with any of these data type

Karim-Mane commented:
> There seem to be a few functions that use data as the first argument and have similar if not the same documentation. If the argument documentation is indeed identical and duplicated I recommend using @inheritParam and only documenting the argument in a single place (probably clean_data() as it's the central exported function).

This is a good point. the `data` argument is used by all the main functions of the cleaning modules. I have opted to document is explicitly in every main function, and made sure to use `@inheritParam` in all helpers and internal functions. Some of these updates were made in commit `326d24b`.

Karim-Mane commented:
> Question: Is there any functionality that uses the report attribute to automate other cleaning functions, or is it just meant for users to inspect? I might be nice to consider if the report can be used by some cleaning functions.

The attribute `report` is primary intended to be used in the `print_report()` function to give users the opportunity to inspect the effects of the cleaning operations. 
However, it is used in several functions to multiple purposes such as:
  * determining how a new element will be added to it depending on existing ones
  * automating the process of identification and selection of a column during cleaning operation if the old name (the name used before the standardize_column_names() function is applied) is provided by a user.

Do you have any suggestion for the usage of this attribute to automate the current or future cleaning operations?

Karim-Mane commented:
> In the DESCRIPTION you're using a development version of {roxygen2}, RoxygenNote: 7.3.2.9000, maybe better to re-render the documentation with a stable release of {roxygen}. I don't think it will cause any issues but just in case as CRAN usually don't like development versions of packages.

This is a good point. I have accounted for this in commit `7651a9d`.

Karim-Mane commented:
> The {cleanepi} package started development in 2022, the copyright in the LICENSE files can be 2022-2025.

Thanks for raising this point. I have account for this in commit `a96c8b6`.

Karim-Mane commented:
> It would be good if you could back-date the NEWS.md and provide some information on previous releases.

This is a good point. I have added some information about previous releases in commit [177eb98](https://github.com/epiverse-trace/cleanepi/pull/207/commits/177eb982423b0976fe442f9205a9dd84b9fcbd5b)

Karim-Mane commented:
> I would remove "curating" from the package Description: in the DESCRIPTION file.

I think the “curating” word is works fine here as all the cleaning modules either helps for inspecting or modifying the values in specific columns, with the ultimate goal of helping the users to curate (when issues are found) and preparing the data for downstream analyses.

Degoot-AM commented on vignettes/cleanepi.Rmd:
```diff
@@ -360,72 +380,75 @@ It also adds two or three elements to the report object:
 
 
-## Standardizing subject IDs
+## Standardizing subject ids
```
```suggestion
## Standardizing subject IDs
```

Degoot-AM commented on vignettes/cleanepi.Rmd:
```diff
@@ -360,72 +380,75 @@ It also adds two or three elements to the report object:
 
 1. **data**: A data frame or linelist (required).
-2. **target_columns**: The name of the column containing the subject IDs in the dataset (required).
+2. **target_columns**: The name of the column containing the subject ids in the dataset (required).
```
```suggestion
2. **target_columns**: The name of the column containing the subject IDs in the dataset (required).
```

Degoot-AM commented on vignettes/cleanepi.Rmd:
```diff
@@ -360,72 +380,75 @@ It also adds two or three elements to the report object:
 The `check_subject_ids()` function returns the input dataset and send a warming when there are some incorrect ids.
 
-In addition to detecting undesirable subject ids, the function will also look for missing and duplicated IDs. As the result of this, the report made from this operation might contain two extra elements: **missing_ids** (a vector of row indexes where there is a missing IDs) and **duplicated_ids** (a data frame of rows with the duplicated IDs). Use the `print_report()` function to display the repor ...
+In addition to detecting undesirable subject ids, the function will also look for missing and duplicated ids. As the result of this, the report made from this operation might contain two extra elements: **missing_ids** (a vector of row indexes where there is a missing ids) and **duplicated_ids** (a data frame of rows with the duplicated ids). Use the `print_report()` function to display the repor ...
```
```suggestion
In addition to detecting undesirable subject IDs, the function will also look for missing and duplicated IDs. As the result of this, the report made from this operation might contain two extra elements: **missing_ids** (a vector of row indexes where there is a missing IDs) and **duplicated_ids** (a data frame of rows with the duplicated IDs). Use the `print_report()` function to display the report made from this operation.
```

Karim-Mane commented:
> Add PR or issue tags to NEWS.md bullet points currently missing them, if possible.

This has been accounted for in commit [59e825a](https://github.com/epiverse-trace/cleanepi/pull/207/commits/59e825a46c60e61193112b2619d02d7ca0501f1e)

Karim-Mane commented:
> I'm a bit confused by the add_to_dictionary() function, I'm not sure how much value this function adds over a user just calling rbind() to add to a dictionary. I've seen the description in the cleanepi.Rmd vignette, but I would appreciate a bit more explanation to where you see this function being used.

* value this function adds over a user just calling rbind()?
      - After identifying misspelled options i.e. other values than those specified in the data dictionary, we propose to the   users the `add_to_dictionary()` function to allow them to explicitly update or add these misspelled values. 
      - Obviously anyone using R would be familiar with functions such as `rbind()` as they are among the basic functionality of the R language. However, in {cleanepi}, it does not matter whether a simple function is used somewhere. We want to be able to diagnose a problem and provide a solution for fixing it. Users are free to use our solution or a customised one.

* a bit more explanation to where you see this function being used
     - the function is intended to offer users with the ability to explicitly update misspelled or add new options in the data dictionary data frame. Another way of doing is allowing for automated misspelled options correction as you suggested in PR #208. This needs to be subjected to multiple examples to evaluate the proportion of unexpected mistakes that could come out of this functionality.

Karim-Mane commented:
> Running the add_to_report() example the output of clean_data() is a <data.table>, I would expect it to be a <data.frame> as that is the input type. I can't see anywhere where {data.table} is called so I assume this is a dependency doing this conversion. Can you check to see if this is also the case for you, and if so it would be good to return the same tabular data type that the user inputs.

The output object from the `clean_data()` function is actually of the same type as the input object. In that example, the input is a `data.table`, so is the output. 
This also consolidate the fact that the functions in the package works fine on data.table objects.

``` r
library(magrittr)
data <- readRDS(system.file("extdata", "test_df.RDS", package = "cleanepi"))
class(data)
#> [1] "data.table" "data.frame"

cleaned_data <- cleanepi::clean_data(
    data = data,
    to_numeric = list(target_columns = "sex", lang = "en"),
    dictionary = NULL
)
#> ℹ Cleaning column names
#> ℹ Removing constant columns and empty rows
#> ℹ Removing duplicated rows
#> ℹ No duplicates were found.
#> ℹ Converting the following  column into numeric: sex
class(cleaned_data)
#> [1] "data.table" "data.frame"

# after cleaning,, the output data has the same type as the input
identical(class(cleaned_data), class(data))
#> [1] TRUE
```

<sup>Created on 2025-02-22 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

Karim-Mane commented:
> Question: in add_to_report() is it possible to add a list element to the report attribute in place rather than extracting it, modifying it and then replacing it. I would only be a marginal gain either way.

* Currently the report object is a <list>. Any cleaning operation that has an output to be added to the report normally adds an extra element to that object. An existing element of this list is only overwritten when the same operation results into a slightly different result than the previous.
* We could create an attribute for every cleaning operation separately instead of having one unique attribute. This is something I am thinking of doing in the coming release.

Karim-Mane commented:
> Would be good to add to the documentation of convert_to_numeric() that it is a wrapper around numberize::numberize()

This is already mentionned in both the package vignette and the design vignette. I have added some text in the function documentation to specify this. See commit [dcf7d85](https://github.com/epiverse-trace/cleanepi/pull/207/commits/dcf7d85fe3a1c1dd9110f26205642f8d624cefad)

Karim-Mane commented:
> test_df1.rds has a typo in the column name, recruted_on_day should be recruited_on_day

Thanks for spotting this out. I have fix the typo in commit [3435485](https://github.com/epiverse-trace/cleanepi/pull/207/commits/34354850a4ff9bdd1e96dba7c168930674e93ee7)

Karim-Mane commented:
> Recently (R 4.3.0) I think the behaviour of as.Date.numeric() changed, it might be worth seeing how this affects convert_numeric_to_date()

This is a good point. I added it as part of the TODO list in the package project.

Karim-Mane commented:
> Question: I'm not sure how much you want to automate because of false positives/negatives but it might be worth having some amount of automated detecting of column types and coercing for convert_*() functions. For example if I have a line list with date columns that are numeric, is it beneficial to not specify the target_columns argument and instead find all numeric columns that could be converted to dates? It would add complexity to the functions, and I'm not saying this is definitely the way to go, but might be something to consider.

Thanks for the suggestion. This is what the `standardize_dates()` function does in the current release version. 
I got feedback from users, including our team members like Hugo, Degoot, and Andree who spotted that the previous behaviour of the function was introducing false positives/negatives. 
Hence, in this release, we opted for the followings:
* stop automatic conversion of numeric into Date from the `standardize_dates()`. Users need to explicitly specify the name of the numeric columns to be converted into Date
* we allow for automatic detection of Date values in `scan_data()`, the function considered as the entry point for a global diagnostic of the input data. In the current development version, the function identifies valid dates among the numeric values using `lubridate::as_date()` and `date_guess()` functions. If valid dates are found, a warning is triggered to alert about the presence numeric values that are potentially of type Date.
     * A numeric value is considered a valid Date if it falls within the range from today's date to 50 years in the past. As this is an arbitrary threshold, we decided not to apply this consideration in `standardize_date()`. It, however, is a good alert system for users to take a close look into the flagged columns.

Karim-Mane commented:
> Question: how does correct_subject_ids() differ from clean_using_dictionary(), they both seem to clean a <data.frame> using a user-specified <data.frame>, albeit with slightly different structures. Could these functions be merged? It could be preferable to channel all <data.frame> based cleaning through clean_using_dictionary()?

* In essence, the `clean_using_dictionary()` is intended for factor columns where a user knows the possible values that a variable can take. It will identify misspelled values and replace the existing ones with their corresponding values from the data dictionary.
* For the subject ids, a user who wants to check whether the IDs comply with the expected format can use the `correct_subject_ids()` function. Here we are not comparing the IDs to a set of pre-defined IDs, but looking for compliance with a specific structure (prefix, suffix, range of numbers, number of characters).
* The `correct_subject_ids()` function allows users to provide replacement for incorrect subject IDs, while making sure that the IDs remain unique. I will explore the possibilities of merging it with the `add_to_dictionary()` function, which allow users to correct for the detected misspelled values.

Karim-Mane commented:
> In check_subject_ids(), it was unclear to me on first reading whether the prefix and suffix arguments are identifying the correct or incorrect ID formats. It would be good to clarify the function documentation to make this clear.

Thanks for the suggestion. I have accounted for this in commit [fdff2b3](https://github.com/epiverse-trace/cleanepi/pull/207/commits/fdff2b3f082ce5a8864a7dee5006f1f52a75040a)

Karim-Mane commented:
> Looking through the different exported function names I find it a bit difficult to identify exactly how the functionality differs depending on the first word in the function name, e.g. clean_*(), standardize_*(), correct_*(), convert_*(). Do these different verbs pertain to different functionality? If not I would give all of these functions the same verb to simplify and unify the function naming. If they do relate to different functionality it would be good to document this in the design principles vignette.

Every function in {cleanepi} is named such that it reflects the action that it does. Functions with a similar effect have a name that starts with the same prefix. The current version of the package contains the following semantic prefixes:

*   `add_` is used to in functions that are made to add something to an existing object
*   `correct_` is used to correct existing values
*   `convert_` is used in the two functions that aim at converting values from one type to another
*   `standardize_` is used to transform some values for compliance with a predefined standard
*   `clean_` is used in `clean_data()` and `clean_using_dictionary()`. It's primary used to define functions that operate on multiple columns.
*   `check_` is used in functions that check whether some values comply with expectations, without transforming them
*   `remove_` is used in functions that remove some rows or columns from the input dataset
    
There are also other functions names that fall outside of this semantic nomenclature due to their the specific nature of their actions: `replace_missing_values()`, `scan_data()`, `find_duplicates()`, `print_report()`, `get_default_params()`

There is room for improvement such as merging `add_to_dictionary()` and `correct_subject_ids()`. I will consider them in the future.

I have added the above to the design vignette in commit [b6baab3](https://github.com/epiverse-trace/cleanepi/pull/207/commits/b6baab336ebd55aae3362f51b9d7ef03cc7dc309)

joshwlambert commented on R/convert_to_numeric.R:
```diff
@@ -4,12 +4,13 @@
 #' function. Furthermore, it identifies columns where the proportion of numeric
 #' values is at least twice the percentage of character values and performs the
-#' conversion in them.
+#' conversion in them. The function internally makes call of the main function
+#' from the {numberize}.
```
Can use `\pkg{numberize}` for formatting.

Karim-Mane replied:
I accounted for this

joshwlambert reviewed (approved):
Thanks for your responses @Karim-Mane. Overall the package looks ready for the next release. Here are a few small comment from re-reviewing the PR.


- I usually only use `<class>` notation for S3 or S4 classes, not types such as `vector` or `list`, for these I use monospace font (backticks). It's personal preference, but just in case you wanted the style to be consistent with the other Epiverse packages.
- I noticed you use the R/latex commands in function documentation (e.g. `\code{}`), Roxygen2 enables you to use markdown, which can be more readable.
- > Do you have any suggestion for the usage of this attribute to automate the current or future cleaning operations?

  Thanks for the examples of where the `report` attribute is used. I don't have any immediate ideas, but as I continue to use the package I'll let you know if I see any potential applications.

- The `NEWS.md` file should go from most recent version at the top to the oldest version at the bottom. As new changes are introduced to the package they are added at the top of the `NEWS` file. See https://r-pkgs.org/other-markdown.html#sec-news for more info.
- The point added to the `NEWS` for `v1.0.0` does not contain much information, could you break down some of the development. If possible linking to PRs or issues is also useful but I acknowledge this may not be possible for the first release if it contains a large amount of development.
- > The output object from the `clean_data()` function is actually of the same type as the input object. In that example, the input is a <data.table>, so is the output.

  Is this to explicitly test whether the function works on `<data.table>`s? If so I would move it to a unit test. Given `<data.table>` is not a package dependency, all examples should use `<data.frame>`s.

Karim-Mane commented:
> Thanks for your responses @Karim-Mane. Overall the package looks ready for the next release. Here are a few small comment from re-reviewing the PR.
> 
>     * I usually only use `<class>` notation for S3 or S4 classes, not types such as `vector` or `list`, for these I use monospace font (backticks). It's personal preference, but just in case you wanted the style to be consistent with the other Epiverse packages.

We decided to generalise this to all data types used in the function documentation. I am happy to discuss this with the wider group to set a standard around and document it in the blueprint for everyone to follow.
> 
>     * I noticed you use the R/latex commands in function documentation (e.g. `\code{}`), Roxygen2 enables you to use markdown, which can be more readable.

I actually make use of both syntaxes in the function documentations. But this again is a style preference that we can agree on with the wider group and get it documented in the blueprint.
> 
>     * > Do you have any suggestion for the usage of this attribute to automate the current or future cleaning operations?
>       
>       
>       Thanks for the examples of where the `report` attribute is used. I don't have any immediate ideas, but as I continue to use the package I'll let you know if I see any potential applications.
> 
>     * The `NEWS.md` file should go from most recent version at the top to the oldest version at the bottom. As new changes are introduced to the package they are added at the top of the `NEWS` file. See https://r-pkgs.org/other-markdown.html#sec-news for more info.

Thanks for the suggestion. I have updated this file in commit [937b042](https://github.com/epiverse-trace/cleanepi/pull/207/commits/937b0421102587b981c1c1c2fefa222c52406bcd)
> 
>     * The point added to the `NEWS` for `v1.0.0` does not contain much information, could you break down some of the development. If possible linking to PRs or issues is also useful but I acknowledge this may not be possible for the first release if it contains a large amount of development.

This is a good point. I only created the NEWS.md file during the first release. I believe that only changes from one version to another are supposed to be documented after the first release.
I have not checked the blueprint on this specific topic, but I think we should add it there to make it explicite that the creation/update of features should be documented from the beginning of the package development.
> 
>     * > The output object from the `clean_data()` function is actually of the same type as the input object. In that example, the input is a <data.table>, so is the output.
>       
>       
>       Is this to explicitly test whether the function works on `<data.table>`s? If so I would move it to a unit test. Given `<data.table>` is not a package dependency, all examples should use `<data.frame>`s.

No - The above was mentioned to emphasise that the package works on all data frame-like objects. Inputs of the package are expected to be either data.frame, linelist or tibble. For simplicity, I have converted this specific object into data.frame in commit [f18be94](https://github.com/epiverse-trace/cleanepi/pull/207/commits/f18be947a4723e073a096b7b1015d577aea438ae)
